// @accounts/common/lib/errors.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var AccountsError = /** @class */ (function (_super) {
    __extends(AccountsError, _super);
    function AccountsError(message, loginInfo, errorCode) {
        var _this = _super.call(this, message) || this;
        Object.setPrototypeOf(_this, AccountsError.prototype);
        _this.epochTime = Date.now();
        if (loginInfo) {
            _this.loginInfo = loginInfo;
        }
        if (errorCode) {
            _this.errorCode = errorCode;
        }
        return _this;
    }
    AccountsError.prototype.toString = function () {
        try {
            return JSON.stringify({
                message: this.message,
                loginInfo: this.loginInfo,
                errorCode: this.errorCode,
                epochTime: this.epochTime,
            });
        }
        catch (e) {
            return ('Error was thrown but could not be serialized. ' +
                'Make sure there is no circular references parameters passed to constructor');
        }
    };
    AccountsError.prototype.serialize = function () {
        return this.toString();
    };
    return AccountsError;
}(Error));
exports.AccountsError = AccountsError;
